File: Problem_Sets_500_599/DC_561_etsy_hard/array_to_binary_tree_arr.js
// This problem was asked by Etsy.

// Given a sorted array, convert it into a height-balanced binary search tree.

// using an array for binary tree representation without a tree node class
function array_to_binary(array) {

    if(array.length <= 2) {
        return array
    }

    let center_index = Math.floor(array.length / 2)
    let left_child_tree = array.slice(0, center_index)
    let right_child_tree = array.slice(center_index + 1)
    
    return [array[center_index],...array_to_binary(left_child_tree),...array_to_binary(right_child_tree)]
}




//↓↓↓   basic test   ↓↓↓

let sorted1 = [2,3,4,5,6,7,8] 
let sorted2 = [0,1,2,3,4,5,6,7,8,9] 

let binary_tree1 = array_to_binary(sorted1)
let binary_tree2 = array_to_binary(sorted2)

console.log(binary_tree1)
console.log(binary_tree2)

File: Problem_Sets_500_599/DC_561_etsy_hard/array_to_binary_tree_arr.py
# This problem was asked by Etsy.

# Given a sorted array / list, convert it into a height-balanced binary search tree.

# using an array for binary tree representation without a tree node class

import math

def list_to_binary(list):
    if len(list) == 0:
        return []
    if len(list) <= 2:
        return list
    else: 
        center_index = math.floor(len(list) / 2)
        left_tree = list[ : center_index]
        right_tree = list[center_index+1 : ]
        return [list[center_index]] + list_to_binary(left_tree) + list_to_binary(right_tree)




# ↓↓↓   basic test   ↓↓↓

sorted0 = [1,2,3]

binary_tree0 = list_to_binary(sorted0)
print(binary_tree0)
#[2,1,3]

sorted1 = [2,3,4,5,6,7,8] 

sorted2 = [0,1,2,3,4,5,6,7,8,9] 

binary_tree1 = list_to_binary(sorted1)
binary_tree2 = list_to_binary(sorted2)

print(binary_tree1)
#[5,3,2,4,7,6,8]


print(binary_tree2)
# [5,2,0,1,3,4,8,6,7,9]